x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/profile/TransportActivateProfileAction.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.action.profile;
+
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.action.GrantApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileAction;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileRequest;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileResponse;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.support.UsernamePasswordToken;
+import org.elasticsearch.xpack.security.authc.AuthenticationService;
+import org.elasticsearch.xpack.security.authc.BearerToken;
+import org.elasticsearch.xpack.security.profile.ProfileService;
+
+public class TransportActivateProfileAction extends HandledTransportAction<ActivateProfileRequest, ActivateProfileResponse> {
+
+    private final ProfileService profileService;
+    private final AuthenticationService authenticationService;
+    private final ThreadContext threadContext;
+
+    @Inject
+    public TransportActivateProfileAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        ProfileService profileService,
+        AuthenticationService authenticationService,
+        ThreadPool threadPool
+    ) {
+        super(ActivateProfileAction.NAME, transportService, actionFilters, ActivateProfileRequest::new);
+        this.profileService = profileService;
+        this.authenticationService = authenticationService;
+        this.threadContext = threadPool.getThreadContext();
+    }
+
+    @Override
+    protected void doExecute(Task task, ActivateProfileRequest request, ActionListener<ActivateProfileResponse> listener) {
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
```

Comment:
I realise this is taken from `TransportGrantApiKeyAction`, but I think you could make use of `SecurityContext#executeWithAuthentication` to simplify it all.

Reply:
I am not entirely sure about using `SecurityContext#executeWithAuthentication`. It is currently only used for secondary authentication where I think it is a better fit because the it means to pretend the secondary authentication is the only authentication object and forget about the initial authentication entirely. 

Grant is very similar. Indeed, it currently has the same practical behaviour once the authentication from the request body is resolved. But I think the semantics is slight different. There were discussions about how the grant type requests could work in the background and one proposal is to serialised both authentication objects. This may not end up to be the solution. But I think it does highlight the difference between grant and secondary authentication in that Grant needs 2 authentiation objects to fully work. So I'd refrain from using `SecurityContext#executeWithAuthentication` right now. 

Also, in [the latest commit](https://github.com/elastic/elasticsearch/pull/82400/commits/59541e88914fd5d44b054ea7c60ae7450bf0fcf0) I extracted a superclass for grant type action to have common `executeWithGrantAuthentication` method. So hopefully this refactoring can still achieve better simplicity without resorting to `SecurityContext#executeWithAuthentication`.